sdk/servicebus/service-bus/test/receiveAndDeleteMode.spec.ts

```diff
@@ -714,3 +722,180 @@ describe("Unsupported features in ReceiveAndDelete mode", function(): void {
     await testRenewLock();
   });*/
 });
+
+describe("Receive Deferred messages in ReceiveAndDelete mode", function(): void {
+  let sequenceNumber: Long;
+
+  async function deferMessage(useSessions?: boolean): Promise<void> {
+    const testMessages = useSessions ? TestMessage.getSessionSample() : TestMessage.getSample();
+    const sender = senderClient.createSender();
+    await sender.send(testMessages);
+    await sender.close();
+    const msgs = await receiver.receiveMessages(1);
+
+    should.equal(Array.isArray(msgs), true, "`ReceivedMessages` is not an array");
+    should.equal(msgs.length, 1, "Unexpected number of messages");
+    should.equal(msgs[0].body, testMessages.body, "MessageBody is different than expected");
+    should.equal(msgs[0].messageId, testMessages.messageId, "MessageId is different than expected");
+    should.equal(msgs[0].deliveryCount, 0, "DeliveryCount is different than expected");
+
+    sequenceNumber = msgs[0].sequenceNumber!;
+    await msgs[0].defer();
+  }
+
+  async function receiveDeferredMessage(): Promise<void> {
+    let deferredMsgs: ServiceBusMessage | undefined;
+
+    deferredMsgs = await receiver.receiveDeferredMessage(sequenceNumber);
+    if (!deferredMsgs) {
+      throw `No message received for sequence number ${sequenceNumber}`;
+    }
+
+    should.equal(deferredMsgs!.deliveryCount, 1, "DeliveryCount is different than expected");
+    await testPeekMsgsLength(receiverClient, 0);
+  }
+
+  /* it("Partitioned Queue: No settlement of the message removes message", async function(): Promise<
+    void
+  > {
+    await beforeEachTest(
+      ClientType.PartitionedQueue,
+      ClientType.PartitionedQueue,
+      undefined,
+      ReceiveMode.peekLock
+    );
+    await deferMessage();
+    await afterEachTest();
+    await beforeEachTest(ClientType.PartitionedQueue, ClientType.PartitionedQueue);
+    await receiveDeferredMessage();
+    await afterEachTest();
+  });
+
+  it("Partitioned Subscription: No settlement of the message removes message", async function(): Promise<
+    void
+  > {
+    await beforeEachTest(
+      ClientType.PartitionedTopic,
+      ClientType.PartitionedSubscription,
+      undefined,
+      ReceiveMode.peekLock
+    );
+    await deferMessage();
+    await afterEachTest();
+    await beforeEachTest(ClientType.PartitionedTopic, ClientType.PartitionedSubscription);
+    await receiveDeferredMessage();
+    await afterEachTest();
+  }); */
+
+  it("Unpartitioned Queue: No settlement of the message removes message", async function(): Promise<
+    void
+  > {
+    await beforeEachTest(
+      ClientType.UnpartitionedQueue,
+      ClientType.UnpartitionedQueue,
+      undefined,
+      ReceiveMode.peekLock
+    );
+    await deferMessage();
+    await afterEachTest();
+    await beforeEachTest(ClientType.UnpartitionedQueue, ClientType.UnpartitionedQueue);
+    await receiveDeferredMessage();
+    await afterEachTest();
```

Comment:
We are essentially creating the sender and receiver clients twice here. Was this so that you could create the receiver in PeekLock mode first, and then again in ReceiveAndDelete mode?

Can we not do the below instead?
- `await beforeEachTest(...)` just like you did 
- `await deferMessage()` just like you did
- `await recevier.close()`
- `receiver = receiverClient.createReceiver(....)
- `await receiveDeferredMessage` just like you did

This may also explain the problem with long running and not exiting tests.
When you defer the message and then re-create the sender/receiver clients, `beforeEachTest` tries to purge the queue/subscription. It enters into a infinite loop there as `peek` finds the deferred message but our `purge` code doesnt purge deferred messages

Reply:
Just tried, not working. More tests are failing than before -

Seeing lot of 
```
Error: The receiver for session "my-session" in "unpartitioned-sessions-topic/Subscriptions/unpartitioned-sessions-subscription" has been closed and can no longer be used. Please create a new receiver using the "createReceiver" function.
```
The tests setup issues are consistently being a bottleneck.

Reply:
If entities are being recreated each time, it's not clear as to why do we still need to purge.

Reply:
> If entities are being recreated each time, it's not clear as to why do we still need to purge.

The purge is there to allow us the option of running tests locally without the tear down and set up of entities before each test. This option is necessary as otherwise we all (and any external contributors who want to contribute) have to set up the Client_Id, Secret, Tenant_Id etc using AAD to run the tests. To reduce the barrier to get tests running, we have this option.

One improvement I can see that we can do here is to skip the purge when `CLEAN_NAMESPACE` is set to `true` i.e. when the entities are being torn-down/set-up. 

> Just tried, not working. More tests are failing than before -
Seeing lot of
`Error: The receiver for session "my-session" in "unpartitioned-sessions-topic/Subscriptions/unpartitioned-sessions-subscription" has been closed and can no longer be used. Please create a new receiver using the "createReceiver" function.`

> The tests setup issues are consistently being a bottleneck.

Those errors have nothing to do with test setup. A couple of minutes into debugging, I found that this is because `isClosed` on session receivers are returning `true` all the time due to a bug that got introduced earlier today. https://github.com/Azure/azure-sdk-for-js/pull/2242 fixes this bug.

Porting that fix in and then following the suggestion in https://github.com/Azure/azure-sdk-for-js/pull/2232#discussion_r275093088 should solve the need to create the clients twice in the same test.

Reply:
The tests are now passing since pulling in latest changes.
Do we still need the tests to be updated based on above comment? @ramya-rao-a

Reply:
The latest changes from master only help in fixing the session receiver related errors that are independent of the changes in this PR.

The initial set of suggested changes in https://github.com/Azure/azure-sdk-for-js/pull/2232#discussion_r275093088 are still applicable as you are still creating the sender/receiver clients twice inside the same test
- I see `purge` resulting in an infinite loop when running tests locally multiple times
- When running this test in CI, it will also fail because `beforeEachTest` will tear down the queue/topic/subscription and create new ones

Reply:
2 out of 8 tests are consistently failing when updated to use the recommended changes.

It seems like there might be a bug in the SDK / Service Bus because other types of entities are passing the tests jus fine. Logged issue to track this - https://github.com/Azure/azure-sdk-for-js/issues/2268

For now, commented out those 2 tests to unblock this issue.

Reply:
Looks like the deferred messages aren't being deleted in the Partitioned Queue/Subscription when receiving in ReceiveAndDelete mode.

This could be a Service side issue as we don't do anything different for partitioned vs unpartitioned entities.

Our next step should be to check the .Net SDK and see if the same operations are successful there.

Reply:
But the tests pass when I do the beforeEachTest() twice, and clean up namespace? Also, it's odd that messages are getting sent twice to these 2 entities. But yes, this can be taken up as part of the new issue investigation.

Reply:
With CLEAN_NAMESPACE set to `true`, the second time `beforeEachTest()` is called, you get a brand new queue/topic/subscription and so the `await testPeekMsgsLength(receiverClient, 0);` must be passing. In this case check what is the output of `receiveDeferredMessage()` call

Reply:
no, I haven't set CLEAN_NAMESPACE = true. Encountering issue where the README instructions were followed but role assignment seems to be missing permissions to delete.
The tests pass without any entities being recreated as such.
I did check the entities as well, there is no message left over as there were otherwise.

Reply:
If CLEAN_NAMESPACE is not set, then new entities will not be created. 
In this scenario, if you were running `beforeEachTest()` twice, then wouldnt the second run get stuck in the infinite loop in `purge`?

Regardless, you have logged the issue #2268 to track this and so that can be used to carry forward the investigation.